allennlp/training/cross_validation.py

```diff
@@ -0,0 +1,317 @@
+import copy
+import logging
+import os
+from typing import Any, Dict, Iterator, Optional, Sequence, Tuple, Union
+
+import sklearn.model_selection
+from numpy.random.mtrand import RandomState
+from overrides import overrides
+from sklearn.model_selection import BaseCrossValidator
+from torch.utils.data import Dataset, Subset
+
+from allennlp.commands.train import TrainModel
+from allennlp.common import Lazy, Registrable, util as common_util
+from allennlp.data import DataLoader, Instance, Vocabulary, DatasetReader
+from allennlp.models.model import Model
+from allennlp.training import Trainer, util as training_util
+from allennlp.training.metrics import Average
+
+
+class CrossValidator(Registrable, BaseCrossValidator):
+    default_implementation = "k_fold"
+
+    def __call__(
+        self, instances: Sequence[Instance]
+    ) -> Iterator[Tuple[Sequence[int], Sequence[int]]]:
+        groups, labels = self._labels_groups(instances)
+        return super().split(instances, labels, groups=groups)
+
+    @overrides
+    def split(self, instances: Sequence[Instance]) -> Iterator[Tuple[Sequence[int], Sequence[int]]]:
+        return self(instances)
+
+    @overrides
+    def get_n_splits(self, instances: Sequence[Instance]) -> int:
+        groups, labels = self._labels_groups(instances)
+        return super().get_n_splits(instances, labels, groups=groups)
+
+    @staticmethod
+    def _labels_groups(
+        instances: Sequence[Instance],
+    ) -> Tuple[Optional[Sequence[Union[str, int]]], Optional[Sequence[Any]]]:
+        labels = [instance["label"] for instance in instances] if "label" in instances[0] else None
```

Comment:
I'm thinking that the key names "label" and "_group" may be defined when the `CrossValidator` object is defined, or be the dataset `Sequence[Instance]`. Thoughts on this? Also, what if it's a nested value (e.g., `instance["metadata"]["_group"]`)?

Reply:
It's not obvious to me what this is doing or why it's needed.  Can you give some more detail?

Reply:
There are 2 options in CV (and when randomly splitting) that are common: stratification and grouping.

The former is about making sure all class labels are represented with a similar balance in train and test. For that, you need to know about the target class (so, we need to know "label").

For the latter, it can be for example about doing leave-one-group-out CV or making sure train and test instances come from different "groups" (whatever groups mean in a given dataset). For that, you need to specify what the group is for each instance.

Reply:
So this works only if the instance have a field called `"label"`?

Reply:
I'd be happier with this if we found a way to let a caller pass in a function to pull out the label or group.

Reply:
> So this works only if the instance have a field called `"label"`?

Yes. Note this is also assumed here (and if IIRC, in more places):

https://github.com/allenai/allennlp/blob/e20997059a6d7db32cdd56df77d710b564349957/allennlp/models/basic_classifier.py#L93

My idea is to have that as a default but to be customizable.

> I'd be happier with this if we found a way to let a caller pass in a function to pull out the label or group.

I think ideally this should come from the `Instance` or the `Dataset`. It seems simpler to be specified by the `CrossValidator` or the function caller. If it's the function caller (`CrossValidateModel`), it can get a function as arg in `__init__`, but how to specify a `Callable` when loading from params (from a json file)? If it's just the function name, what if it's more complex, like a metadata field?

Reply:
One option is to give a field name and have a function that pull that field out of the instance.  You could also take that a little bit farther, to handle metadata fields, or something.

Another option is to introduce an abstraction over these functions, which you can then create `FromParams`.  That seems way too heavy for this problem, though.

Reply:
I changed this. It just accepts a key name from params, but more flexible in the constructor.

As a separate issue, I was thinking that it'd be good for `Callable`s to be registered. They can be instantiated by creating a class that defines `__call__` (note that a class `C` that does so satisfies `isinstance(C, Callable)`). So, for example, here there could be a class that gets instantiated when a `Callable` arg gets assigned to a type like `instance_field_by_key` from params. Similarly for other `Callable` args.

Related, what about allowing any class to have a registry, not only `Registrable` subclasses? Instead of doing `Subclass.register(...)` to do like `Registrable.register(Class, ...)`. This can help `Callable`, and also maybe avoid intermediate classes such as `Optimizer`.

Reply:
It only really makes sense to register an abstract class if its subclasses are interchangeable when used downstream.  That's not true for `Callable`, because the arguments and return types of different `Callables` don't have to be the same.  So if you find yourself tempted to register something as general as a `Callable`, you should think about adding a more useful abstraction, or finding a different way to do what you want without adding so heavy as a new abstraction.

As for `Optimizer`, the main reason that exists is only so that we can use the type annotation in a constructor.  Doing `Registrable.register([something? "optimizer"?], ...)` defeats the entire purpose, because what do we put for the type annotation on the argument to the trainer?

Reply:
> As for Optimizer, the main reason that exists is only so that we can use the type annotation in a constructor.

But we could, in theory, use PyTorch's `Optimizer` if we needn't subclass `Registrable`, right?

> because what do we put for the type annotation on the argument to the trainer?

We still put `Optimizer`. It'd be like `Registrable.register(Optimizer, ...)`.

Reply:
That was essentially what we did before I removed all of our custom `from_params` logic.  It wasn't exactly this, but it was close enough.  The trouble with that is that you have to do crazy magic in `from_params` that makes everything hard to use, because the objects we're trying to construct don't use `FromParams`, and don't have type annotations.

Reply:
I see. If they had type annotations then it'd be easy? I think it'd be cool to add all subclasses of a certain type, supposing they have type annotations (e.g., all PyTorch optimizers under `Optimizer`, setting a default one, and setting names based on the class names).

Reply:
If they had type annotations on their argument, we'd have to change our `FromParams` logic a bit, but in principle we'd be able to accommodate constructing them using just `torch.nn.Optimizer` as a type annotation.